### DISCUSSION TITLE
Places glitches on mobile (150k photos)

### DISCUSSION THREAD

**Discussion by stefanocolafranceschi** [photoprism/photoprism#4018] [Q&A]
Hi all,
The places map on any mobile any browser causes the web app to restart. Tested on android/iOS latest and FF, safari, chrome, opera. From any computer no problem, the issue occurs when the number of picture is relatively large (150k) and from a mobile device. The behavior seems like the page timeouts or overloads the mobile browser, is there a hidden setting we could play on the server to handle this?
I am running latest photoprism docker on Ubuntu (dedicated a disc just for photoprism cache so there is space for that).

**Comment by lastzero** [comment]
That seems possible, although I hear this the first time. We've tested it with 100k pictures and it worked well. What specific devices (brand and model) did you test with?

**Reply by stefanocolafranceschi** [reply]
Tested on iPhone SE and Motorola g pure, I was looking into browser settings but on mobile apps but they have limited options and debugging features.

Perhaps once the photoprism web server (nginx?) detects mobile, some compression and/or longer timeout allowed?

**Reply by lastzero** [reply]
iPhone SE latest or e.g. SE 2?

**Reply by lastzero** [reply]
If you don't have gzip compression enabled in PhotoPrism you should do so as that makes a major difference...

**Reply by stefanocolafranceschi** [reply]
I had gzip enabled already (it was default), on iPhone SE latest, any other settings I could look into?

**Reply by lastzero** [reply]
Did you open it using a local network connection or a mobile network? If you have a Mac, you can check the mobile browser logs for helpful information:
- https://developer.apple.com/documentation/safari-developer-tools/inspecting-ios

We are happy to test it (again) with a larger library, but don't have one available at the moment that would have that many pictures with GPS data. Since the clustering takes place in the user interface, there is a chance that some optimization can be implemented. If it is related to the response size as such, we need to rework the API to return the cluster sizes only at low zoom levels and the pictures only when zooming in.

Any help from other users to test and improve this will be much appreciated! :+1:

**Reply by stefanocolafranceschi** [reply]
In the web inspector, I do not see any error happening, but in the network tab I do see that the following takes quite a while to load (tens of seconds) and while the load keeps going in the background meanwhile the "Places" page stops loading. (It's a 50mb json ;-)
```
Summary
URL: http://10.3.61.171/api/v1/geo?count=500000&offset=0&q=&s=&public=true&quality=3
Status: 200 OK
Source: Network
Initiator: 
app.6af78ad436ea71dcf21b.js:2:1573687

Request
Accept: application/json, text/plain, */*
Accept-Language: en
Referer: http://10.3.61.171/library/places
User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 17_2_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.2 Mobile/15E148 Safari/604.1
X-Client-Uri: /static/build/app.6af78ad436ea71dcf21b.js
X-Client-Version: 231128-f48ff16ef-Linux-AMD64-Plus
X-Session-ID: caadf5d5f97fcd7c60937c4162fe6832107cf299a1d64499

Response
Content-Security-Policy: frame-ancestors 'none';
Content-Type: application/json
Date: Fri, 19 Jan 2024 19:16:52 GMT
Transfer-Encoding: Identity
X-Count: 166391
X-Download-Token: 3ustypb3
X-Frame-Options: DENY
X-Limit: 500000
X-Offset: 0
X-Preview-Token: 3aucqqwj

Query String Parameters
count: 500000
offset: 0
q
s
public: true
quality: 3
```

I thought perhaps you could generate pictures to test on your end, the following will generate 50k pictures (works/tested on my mac using ImageMagick and exiftool)
```
for i in {1..50000}; do
  convert -size 1000x1000 xc:gray +noise gaussian jpeg: | \
    convert -define jpeg:extent=1048576 - \
      "random_image_with_gps_$i.jpg"

  exiftool -gpslatitude=$(awk 'BEGIN {srand(); srand(); print rand() * 180 - 90}') \
            -gpslongitude=$(awk 'BEGIN {srand(); srand(); print rand() * 360 - 180}') \
            -gpsaltitude=$(awk 'BEGIN {srand(); srand(); print rand() * 10000}') \
            "random_image_with_gps_$i.jpg"
done
```

just greyish images, but they will work for the purpose of having a large database, with random GPS data (around 780kb each picture)

p.s. yes this happens in the same lan network.

**Reply by lastzero** [reply]
The "identity" transfer encoding was removed in RFC 7230:
https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.1

Can you try connecting directly to your PhotoPrism instance with your mobile device to see if that makes a difference?

**Comment by stefanocolafranceschi** [comment]
what do you mean? The previous log was referred to when I connected directly to photoprism from my mobile StefanoCOn Jan 19, 2024, at 3:14 PM, Michael Mayer ***@***.***> wrote:﻿
The "identity" transfer encoding was removed in RFC 7230:
https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.1
Can you try connecting directly to your PhotoPrism instance with your mobile device to see if that makes a difference?

**Reply by lastzero** [reply]
That's interesting, because I've never seen that header value in a browser console, and there's no indication that the client (the phone) accepts gzip or that the server (PhotoPrism) would send compressed data unless you left out some of the headers sent and received?

**Comment by stefanocolafranceschi** [comment]
I just posted the network request from geo, there are many other requests, geo seemed the one that did not resolve in milliseconds and with a relatively large payload (megabytes). I could look and compare Safari running on my MacBook vs Safari running on my iPhone and see if the headers kinda match.. any other things I could look into? Or tests?Also when I had much less picture Places worked, that’s why the larger payload seemed to me interesting to study, what do you think?StefanoCOn Jan 19, 2024, at 3:41 PM, Michael Mayer ***@***.***> wrote:﻿
That's interesting, because I've never seen that header value in a browser console, and there's no indication that the client (the phone) accepts gzip or that the server (PhotoPrism) would send compressed data unless you left out some of the headers sent and received?

**Reply by lastzero** [reply]
The request should look something like this, note the `Accept-Encoding` header:

```
GET /api/v1/geo?count=500000&offset=0&q=&s=&public=true HTTP/1.1
Accept: application/json, text/plain, */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en
Cache-Control: no-cache
```

Since mobile devices in particular want to reduce the required bandwidth, I assume that they don't intentionally omit it in order to maximize the download size? :thinking:

**Reply by lastzero** [reply]
Please also let us know which version of PhotoPrism you are using, as we do not test with older versions by default.

**Reply by lastzero** [reply]
I changed the position of the compression handler so that we can test if this makes a difference: https://github.com/photoprism/photoprism/commit/d98a1cb2ab1a237f926c6c0e57c090bebb18c3ba

**Reply by stefanocolafranceschi** [reply]
latest 3.5

on windows10 (chrome) the headers on my end seem this "general" and "request Headers":
```
Request URL:
http://10.3.61.171/api/v1/geo?count=500000&offset=0&q=&s=&public=true&quality=3
Request Method:
GET
Status Code:
200 OK
Remote Address:
10.3.61.171:80
Referrer Policy:
strict-origin-when-cross-origin

```

```
Accept:
application/json, text/plain, */*
Accept-Language:
en
Dnt:
1
Referer:
http://10.3.61.171/library/places
User-Agent:
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36
X-Client-Uri:
/static/build/app.6af78ad436ea71dcf21b.js
X-Client-Version:
231128-f48ff16ef-Linux-AMD64-Plus
X-Session-Id:
1760aefb604c4c435374228e6fa15395345bbb7d3961ad2c
```

on Chrome (win10) Places works, although I do not see the gzip feature, is that normal?

**Comment by stefanocolafranceschi** [comment]
Ok I would love to try but my installation runs on docker, is there a way to quickly patch it?StefanoCOn Jan 19, 2024, at 4:50 PM, Michael Mayer ***@***.***> wrote:﻿
I changed the position of the compression handler so that we can test if this makes a difference: d98a1cb

**Reply by lastzero** [reply]
I've quickly patched it: You can use the develop branch or the `photoprism/photoprism:test` image available on Docker Hub.

**Reply by stefanocolafranceschi** [reply]
Well.. the "request header" is not different! It does show the gzip!

```
Accept:
application/json, text/plain, */*
Accept-Encoding:
gzip, deflate, br
Accept-Language:
en
Connection:
keep-alive
Dnt:
1
Host:
localhost:2342
Referer:
http://localhost:2342/library/places
Sec-Ch-Ua:
"Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"
Sec-Ch-Ua-Mobile:
?0
Sec-Ch-Ua-Platform:
"Windows"
Sec-Fetch-Dest:
empty
Sec-Fetch-Mode:
cors
Sec-Fetch-Site:
same-origin
User-Agent:
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36
X-Client-Uri:
/static/build/app.6af78ad436ea71dcf21b.js
X-Client-Version:
231128-f48ff16ef-Linux-AMD64-Plus
X-Session-Id:
0ce3d443a158185d0b4f44449dc84f8abbc565402e1ba31f
```

I run your test image on a test machine, the ultimate proof would be I deploy your test on my "production" machine that contains all the photos, to check that the gzip feature really does the job. Before I do that, is there anything else in your test image that one has to be aware of?

p.s. also to index my 200k photos it took so much time, if I deploy your test image, is there a recipe to keep my database or do I need to reindex?

**Reply by lastzero** [reply]
@stefanocolafranceschi  The current test image should be generally safe to use (I also use it for my personal instance). However, we are currently adding more authentication options, so users will have to log in again after the update. You may also need to run the `photoprism auth reset` command in a terminal when upgrading to the stable release. However, this has nothing to do with your pictures, so the only side effect may be that you have to log in again and/or manually reset the existing user sessions to force a re-login. I'll be happy to help should you run into any issues while testing and/or updating!

**Reply by lastzero** [reply]
Our `photoprism/photoprism:test` image has just been updated with additional changes and improvements related to the caching and compression headers, so any remaining issues should be fixed. Thanks for letting us know! I'm not sure yet how and when this broke, as we of course tested the gzip compression before we released it (and many times after).

**Reply by stefanocolafranceschi** [reply]
Ok I pulled the test image run it on my storage thing, at this time I see the gzip everywhere but when I try on mobile (iPhone) still observe the same behaviour; on any computer "Places" works as before, actually a bit faster but the problem on the mobile is still there ;-(

**Reply by lastzero** [reply]
Thanks for testing it! How's the response size compared to as it was before? If that wasn't the problem, we may have to change the API endpoint to return just the clusters when there are many results. As a temporary fix, the maximum number of results could be reduced. But then you also won't see them on a desktop browser.

**Comment by stefanocolafranceschi** [comment]
before it was 50MB now it is 8MB, in fact it loads faster. I wonder why the
client (from mobile) can't wait whatever time until the json is returned.
Also, I understand you return the "entire map thing", I tried to zoom the
map here and there and it nicely work offline, once those MB are loaded (on
a computer) I see no additional network requests (until at certain zoom);
would it work to load only what's barely needed to populate the map rather
than the entire map (and then reload when needed aka the user would zoom)?

StefanoC

**Reply by lastzero** [reply]
The search results for the map are already heavily reduced to lower the response size and load time. Since the clustering is currently performed by the map library in the UI (based on a demo I saw where the developers said that millions of coordinates are not a problem...), we can't just send less, because then the number of pictures displayed would not match, resulting in heaps of bug reports...

**Reply by lastzero** [reply]
Maybe a phone with more memory could make a difference - would you be able to test that somehow? It could also make a difference if the server is trusted. So a valid HTTPS certificate might help (if you haven't tested that yet), and if you install the app on your home screen, it might also be allowed to use more resources!?

**Comment by stefanocolafranceschi** [comment]
Is gzip set to maximum compression? Reading online I hear that Brotli might work better than gzip?Can the large json be split in chunks so the mobile client could handle that?StefanoCOn Jan 20, 2024, at 2:28 PM, Michael Mayer ***@***.***> wrote:﻿
The search results for the map are already heavily reduced to lower the response size and load time. Since the clustering is currently performed by the map library in the UI (based on a demo I saw where the developers said that millions of coordinates are not a problem...), we can't just send less, because then the number of pictures displayed would not match, resulting in heaps of bug reports...

**Reply by lastzero** [reply]
See above for solutions / workarounds I would try first. I currently assume this is a cpu or memory limitation on your mobile devices, so it only depends on the total amount of data and cannot be solved by loading the same data in multiple batches.

**Comment by stefanocolafranceschi** [comment]
I had tried the app on the phone screen and the “request desktop site” but it did not help; why the certificate would help?I will try a newer phone although I thought to have a decent one :-oI agree that some resources get saturated, that’s why I was thinking if the large JSON can be streamed and processed incrementally, that might avoid the crash.StefanoCOn Jan 20, 2024, at 3:15 PM, Michael Mayer ***@***.***> wrote:﻿
See above for solutions / workarounds I would try first. I currently assume this is a cpu or memory limitation on your mobile devices, so it only depends on the total amount of data and cannot be solved by loading the same data in multiple batches.

**Reply by lastzero** [reply]
If you google for it, you'll find a bunch of memory and Web GL related bug reports for Safari on iOS (both used by the map), for example:
- https://github.com/emscripten-core/emscripten/issues/19374#issuecomment-1576848289

I'm on my phone right now, so I can't get a definite answer on which resources exactly are dependent on the level of trust a website has and even if I could get you numbers for a specific browser version then another release could have other limits...

**Comment by stefanocolafranceschi** [comment]
Yes you’re right, https plays a role, I now run photoprism with a self signed certificate, I still get a warning but that should not matter. On my iPhone I still experience the glitch in “Places”, I tested on Chrome and Safari (on the computer no problem).StefanoCOn Jan 20, 2024, at 3:45 PM, Michael Mayer ***@***.***> wrote:﻿
If you google for it, you'll find a bunch of memory and Web GL related bug reports for Safari on iOS (both used by the map). I'm on my phone right now, so I can't get a definite answer on which resources exact are dependent on the level of trust a website has and even if could get you numbers for a specific browser version than another release could have other limits...
